kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java

```diff
@@ -228,7 +229,94 @@ public Alert apply(String s) {
+                        )));
+  }
+
+  @Test
+  public void testKafkaSinkAndAlertsPublisherWithNullKey() throws Exception {
+    Schema schema = Schema.recordOf(
+      "user",
+      Schema.Field.of("id", Schema.nullableOf(Schema.of(Schema.Type.LONG))),
+      Schema.Field.of("first", Schema.of(Schema.Type.STRING)),
+      Schema.Field.of("last", Schema.of(Schema.Type.NULL)));
+
+    // create the pipeline config
+    String inputName = "sinkTestInput";
+
+    String usersTopic = "records";
+    String alertsTopic = "alerts";
+    Map<String, String> sinkProperties = new HashMap<>();
+    sinkProperties.put("brokers", "localhost:" + kafkaPort);
+    sinkProperties.put("referenceName", "kafkaTest");
+    sinkProperties.put("topic", usersTopic);
+    sinkProperties.put("schema", schema.toString());
+    sinkProperties.put("format", "csv");
+    sinkProperties.put("key", "last");
+    sinkProperties.put("async", "FALSE");
+    sinkProperties.put("compressionType", "none");
+
+    Map<String, String> alertProperties = new HashMap<>();
+    alertProperties.put("brokers", "localhost:" + kafkaPort);
+    alertProperties.put("topic", alertsTopic);
+
+    ETLStage source = new ETLStage("source", MockSource.getPlugin(inputName));
+    ETLStage sink =
+      new ETLStage("sink", new ETLPlugin("Kafka", KafkaBatchSink.PLUGIN_TYPE, sinkProperties, null));
+    ETLStage transform = new ETLStage("nullAlert", NullAlertTransform.getPlugin("id"));
+    ETLStage alert =
+      new ETLStage("alert", new ETLPlugin("KafkaAlerts", KafkaAlertPublisher.PLUGIN_TYPE, alertProperties));
+
+    ETLBatchConfig pipelineConfig = ETLBatchConfig.builder("* * * * *")
+      .addStage(source)
+      .addStage(transform)
+      .addStage(sink)
+      .addStage(alert)
+      .addConnection(source.getName(), transform.getName())
+      .addConnection(transform.getName(), sink.getName())
+      .addConnection(transform.getName(), alert.getName())
+      .build();
+
+    // create the pipeline
+    ApplicationId pipelineId = NamespaceId.DEFAULT.app("testKafkaSink");
+    ApplicationManager appManager = deployApplication(pipelineId, new AppRequest<>(APP_ARTIFACT, pipelineConfig));
+
+
+    Set<String> expected = ImmutableSet.of("100,samuel,jackson",
+                                           "200,dwayne,johnson",
+                                           "300,christopher,walken",
+                                           "400,donald,trump");
+
+    List<StructuredRecord> records = new ArrayList<>();
+    for (String e : expected) {
+      String[] splits = e.split(",");
+      StructuredRecord record =
+        StructuredRecord.builder(schema)
+          .set("id", Long.parseLong(splits[0]))
+          .set("first", splits[1])
+          .set("last", splits[2])
+          .build();
+      records.add(record);
+    }
+
+    // Add a null record to get an alert
+    StructuredRecord nullRecord =
+      StructuredRecord.builder(schema)
+        //.set("id", 500L)
```

Comment:
Why is the id not set here ? The key field is `last` and it is set to null.

Reply:
Uncomment this code. `id` is set now. b6aa798cc20abacecff2b7d570e47d1678017c88
Yes, the key field is `last` and it is set to `null`